esp-wifi/src/compat/common.rs

```diff
@@ -47,46 +47,78 @@ impl RawQueue {
         }
     }
 
-    fn free_storage(&mut self) {
+    pub(crate) fn release_storage(&mut self) {
         unsafe {
             free(self.storage);
         }
         self.storage = core::ptr::null_mut();
     }
 
-    fn enqueue(&mut self, item: *mut c_void) -> i32 {
-        if self.count() < self.count {
-            unsafe {
-                let p = self.storage.byte_add(self.item_size * self.current_write);
-                p.copy_from(item as *mut u8, self.item_size);
-                self.current_write = (self.current_write + 1) % self.count;
+    pub(crate) fn enqueue(&mut self, item: *mut c_void) -> i32 {
+        critical_section::with(|_| {
```

Comment:
`self` is mutable, why do we need a critical section?

Reply:
Because we use access this globally from different contexts iirc. It would be nice if we could make this (or use a an existing queue) that is lock-free.

Reply:
I think your comment makes this code UB.

Reply:
Or maybe not, but this is questionable at best, the struct should live in a Mutex<RefCell> and locking should be on the outside.

Reply:
Yeah more than likely, two mutable references to `self` if any queue API is called.

We need to change these to `&self` and have interior mutability, or when we create the queues we also wrap them in a `Mutex<RefCell<>>` externally.